src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs

```diff
@@ -481,13 +544,102 @@ private void DeletePack(PackInfo packInfo)
+        private string GetManifestInstallRecordPath(ManifestId manifestId, ManifestVersion manifestVersion, SdkFeatureBand featureBand, SdkFeatureBand referencingFeatureBand) =>
+            Path.Combine(_workloadMetadataDir, InstalledManifestsDir, "v1", manifestId.ToString(), manifestVersion.ToString(), featureBand.ToString(), referencingFeatureBand.ToString());
+
+        void WriteManifestInstallationRecord(ManifestId manifestId, ManifestVersion manifestVersion, SdkFeatureBand featureBand, SdkFeatureBand referencingFeatureBand)
+        {
+            var path = GetManifestInstallRecordPath(manifestId, manifestVersion, featureBand, referencingFeatureBand);
+            if (!Directory.Exists(Path.GetDirectoryName(path)))
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(path));
+            }
+            using var _ = File.Create(path);
+        }
+
+        private Dictionary<(ManifestId manifestId, ManifestVersion manifestVersion, SdkFeatureBand manifestFeatureBand), List<SdkFeatureBand>> GetAllManifestInstallRecords()
+        {
+            Dictionary<(ManifestId manifestId, ManifestVersion manifestVersion, SdkFeatureBand manifestFeatureBand), List<SdkFeatureBand>> ret = new();
+
+            var installedManifestsDir = Path.Combine(_workloadMetadataDir, InstalledManifestsDir, "v1");
+
+            if (!Directory.Exists(installedManifestsDir))
+            {
+                return ret;
+            }
+
+            foreach (var manifestIdDir in Directory.GetDirectories(installedManifestsDir))
+            {
+                var manifestId = new ManifestId(Path.GetFileName(manifestIdDir));
+                foreach (var manifestVersionDir in Directory.GetDirectories(manifestIdDir))
+                {
+                    var manifestVersion = new ManifestVersion(Path.GetFileName(manifestVersionDir));
+                    foreach (var manifestFeatureBandDir in Directory.GetDirectories(manifestVersionDir))
+                    {
+                        var manifestFeatureBand = new SdkFeatureBand(Path.GetFileName(manifestFeatureBandDir));
+                        foreach (var featureBandInstallationRecord in Directory.GetFileSystemEntries(manifestFeatureBandDir))
+                        {
+                            var referencingFeatureBand = new SdkFeatureBand(Path.GetFileName(featureBandInstallationRecord));
+
+                            if (!ret.TryGetValue((manifestId, manifestVersion, manifestFeatureBand), out var referencingFeatureBands))
+                            {
+                                referencingFeatureBands = new List<SdkFeatureBand>();
+                                ret[(manifestId, manifestVersion, manifestFeatureBand)] = referencingFeatureBands;
+                            }
+
+                            referencingFeatureBands.Add(referencingFeatureBand);
+                        }
+                    }
+                }
+            }
```

Comment:
Are the exceptions thrown here okay if the folder structure doesn't match what is expected?

Reply:
Good point, it would probably be better to include the full path if there's something that doesn't parse correctly.